File: src/lib/column-keys.ts
/** Separator between table alias and column name in selected-column keys. */
export const COLUMN_KEY_SEP = '\0';

export function columnKey(alias: string, column: string): string {
	return `${alias}${COLUMN_KEY_SEP}${column}`;
}

export function parseColumnKey(key: string): { alias: string; column: string } {
	const sep = key.indexOf(COLUMN_KEY_SEP);
	if (sep !== -1) {
		return { alias: key.slice(0, sep), column: key.slice(sep + 1) };
	}
	// Legacy keys used a single dot between alias and column.
	const dot = key.indexOf('.');
	if (dot === -1) {
		return { alias: key, column: '' };
	}
	return { alias: key.slice(0, dot), column: key.slice(dot + 1) };
}

export function columnKeyPrefix(alias: string): string {
	return `${alias}${COLUMN_KEY_SEP}`;
}

File: compose-app/shared/src/main/kotlin/com/safedb/query/ColumnKeys.kt
package com.safedb.query

/** Separator between table alias and column name in selected-column keys. */
const val COLUMN_KEY_SEP = '\u0000'

fun columnKey(alias: String, column: String): String = "$alias$COLUMN_KEY_SEP$column"

fun columnKeyPrefix(alias: String): String = "$alias$COLUMN_KEY_SEP"

fun parseColumnKey(key: String): Pair<String, String> {
    val sep = key.indexOf(COLUMN_KEY_SEP)
    if (sep != -1) {
        return key.substring(0, sep) to key.substring(sep + 1)
    }
    // Legacy keys used a single dot between alias and column.
    val dot = key.indexOf('.')
    if (dot == -1) {
        return key to ""
    }
    return key.substring(0, dot) to key.substring(dot + 1)
}
